### ISSUE TITLE
Scripts loaded twice for some themes

### ISSUE THREAD

**Issue by jhabdas** [comfusion/hyperdrive#16]
As mentioned in https://github.com/wp-id/hyperdrive/issues/7#issuecomment-298815216:

> the plugin will [...] bug with the drop down menu of the free [poseidon theme](https://themezee.com/themes/poseidon/)

I looked at the _free_ version of the Poseidon theme and it's built to drop scripts into the document `head`. It does this by enqueuing scripts in its `poseidon_scripts()` method in `functions.php` without using the [`$in_footer` parameter](https://developer.wordpress.org/reference/functions/wp_enqueue_script/#parameters) of `wp_enqueue_script`.

When [YSlow](https://github.com/marcelduran/yslow/) was introduced years ago by _Yahoo!_ it became a Web best practice to load blocking scripts such as those used by WP Core (jQuery) and Poseidon near the bottom of the `body` to allow page content to become visible before all external scripts completed loading and executing.

As it stands Poseidon (or just about any theme for that matter) shouldn't be loading external scripts in the head of the document as doing so is not performant (especially on mobile) and WordPress has made it reasonably simple to achieve greater speed as things are using $in_footer.

Hyperdrive does not currently try and optimize scripts loaded in the document `head` and, though it's possible (as I show through example below), I'm wary of doing so for the following reasons:

- Attempting to override load order of scripts loaded in the HEAD of a document could lead to unexpected issues for `async` scripts loaded early (like Google Analytics), which we should not be concerned about as they're already being loaded asynchronously and typically do not contain dependencies.
- Doing so would cause us to have to set a _priority_ on the plugin load order, and that would be an anti-pattern as we'd be competing for priority with an unknown number of themes and plugins which try and set their own priority.
- With or without use of this plugin, theme authors should be encouraged to move synchronous (blocking) scripts making external calls outside the `head` of their document and preferably near the bottom of the page whenever possible.
- Hyperdrive shouldn't become an enabler for themes not following established best practices as not everyone will be using Hyperdrive and, as a result, encouraging others to improve theme/plugin performance will have a greater benefit for the vast majority of end users.
- Lastly, thought it's not in WP core yet to my knowledge, they may add Native support for traditional async scripting in the future, and we shouldn't stand in the way of that work.

That said, we could always expose this option to power users later via the plugin admin UI if it makes sense.

---

**Outcomes**

As a result of this investigation the following things should happen:

- [x] We add notice themes loading external scripts without using established best practices for Web performance _may_ not be fully supported by Hyperdrive. /cc @ariffsetiawan 
- [x] Additionally, we should encourage users to reach out to theme or plugin maintainers and request they optimize their scripts. /cc @ariffsetiawan 
- [x] Open an issue to create a theme/site analyzer to check how much of a performance gain can be had with Hyperdrive (a great marketing tool, as well).
- [x] Consider 1.0.0 duplicate script loading remediation steps for non-standard themes pre-launch

---

I'm leaving this issue open until the above two items are complete.

Also, I ran some tests to see what kind of performance gain the free version of Poseidon would have if it were running with Hyperdrive. The before and after can be found below. Focus on the `DOMContentLoaded` event. That's an approximation of how long it takes a user to see content when loaded over a _Good 2G_ connection (and for illustrative purposes in this test):

Poseidon **without Hyperdrive** (scripts loaded in `head`):

<img width="644" alt="screen shot 2017-05-05 at 5 55 44 pm" src="https://cloud.githubusercontent.com/assets/440298/25743206/c899cb7e-31c5-11e7-945e-873df73e5353.png">

- DOMContentLoaded: ~4 seconds over Good 2G

Poseidon **with Hyperdrive** (using the `1.0.0-beta` customized):

<img width="644" alt="screen shot 2017-05-05 at 5 56 09 pm" src="https://cloud.githubusercontent.com/assets/440298/25743296/2e34f1ca-31c6-11e7-9a8f-d83a172c35cc.png">

- DOMContentLoaded: < 2 seconds over Good 2G

The above results can be reproduce by customizing Hyperdrive `engage` to hook into `wp_head` by adding a priority of 8 or less (the antipattern described above) and using Chrome Dev Tools with caching disabled and speed throttling. After doing this Poseidon menus had no issue, so I consider this a theme concern.

**Comment by jhabdas** [comment]
I have left a support request on the forums of the Poseidon theme requesting the change described here and linked them back to this issue: https://wordpress.org/support/topic/move-script-loading-to-footer/.

/cc @ariffsetiawan @nielslange @nickl1

**Comment by ariffsetiawan** [comment]
@jhabdas already updated docs content

**Comment by jhabdas** [comment]
>  Consider 1.0.0 duplicate script loading remediation steps for non-standard themes pre-launch

Will be effectively solved with #24 so I'm calling the quoted task done and closing this issue.

**Comment by jhabdas** [comment]
Reopening for continued conversation in light of #26.

**Comment by EiraStudios** [comment]
I've tested this on several custom themes and Twenty Seventeen & Poseidon and while PR #24 fixed scripts being loaded twice via `fetchInject()`, it didn't resolve the issue with scripts not being dequeued and therefore still included in the header as normal. This obviously led to them being loaded twice, once via `fetchInject()` and again via the normal includes.

Essentially without PR #24 there were a chance scripts could be loaded three times, in my primary test case there were four scripts loaded three times; twice using `fetchInject()` and once the normal way.

`calibrate_thrusters()` doesn't seem to successfully dequeue the scripts. It looks like it's because these are being enqueued after Hyperdrive hooks into `wp_head` so to fix this we would need to hook into `wp_print_scripts`, which is the last hook called before the scripts are rendered to the header.

Switching to this hook fixes the scripts not being dequeued but then another issue crops up with jQuery Migrate always finishes loading before jQuery Core does which results in _`"jQuery not defined error"`_.

I've fixed both issues and a couple of other changes in PR: #27

**Comment by jhabdas** [comment]
@EiraStudios thanks for the info. If you have a chance, please review my notes in the opening comment of this issue for concerns on aggressively dequeuing head scripts. Let's say we did have a standard way to pass over existing "type=async" scripts in the head, we'd still not be doing justice by pushing for Web standards in the myriad of free themes out there still for some reason not loading scripts in the footer.

AFAICT WordPress does not natively support any method for enqueuing standard "async" scripts, and that's the place to start (will require Core issue if not already one at present). If they do add this, that'll help for scripts without dependencies, but doesn't reverse the number of DOM hacks existing at current as a rusult of the Core functionality to do so, which IMO should've been added at least 2 years ago.

Beyond that, and until I see more research on more themes and scripting anti-patterns currently in place out in the wild, my suggestion is we only provide a means by which themes loading scripts in the are encouraged to move their scripts into the footer (hence the theme analyzer issue).

Worth noting is I don't see how at current, without hacking DOM - out of the question for this plugin - we could do any good to by trying to correct already poorly coded themes. Because if we did those theme issues already lying in wait, would result in a flood of support issues this plugin is not, nor should be, handling.

That said, the takeaways are:

- Theme analyser according to plan (because themes want to support Hyperdrive, and not the other way around)
- And Core issue for natively enqueuing async scripts in the head (where most of the current hacks are).

Lastly, wp_print_scripts() is used outside what WP dubs the "front end". If we were doing anything with scripts enqueued in the head, it should be done with wp_print_head_scripts(). But only after the theme analyzer is built out.

**Comment by EiraStudios** [comment]
@jhabdas Ah, I've just had a quick read of your notes, and it seems I have misunderstood what the plugin was trying to do.

_Sorry about the length of this message, it turned out to be quite long, so I've split it up into sections to make it easier to follow hopefully._

## Plugin Behaviour
> Hyperdrive does not currently try and optimize scripts loaded in the document `head`

This isn't what I've been seeing in my tests (which is what's led to the confusion on my part 😆). As a test, I've setup a site that has about 18 scripts loaded in total (including plugin scripts). I removed three of the theme scripts from being enqueued into the footer so that they'd enqueue to the `head`. 

What I've observed happening is the scripts that I've switched to load in the `head` are still being added to the `fetchInject()` array, however, they're not being dequeued from the `head`, resulting in them being loaded twice. 

It's worth noting here that jQuery (Core & Migrate) included by WordPress always seems to be added the `head` and as such are always loaded twice, due to them not being dequeued.

The only way I was able to get Hyperdrive to dequeue all the scripts in the `fetchInject()` array while being hooked into `wp_head` was by giving the hook a priority of one.

Hyperdrive does, however, successfully dequeue the footer scripts present in `fetchInject()`, and on some occasions, some of the scripts in the `head` are dequeued while being hooked into `wp_head` without a priority set. It just depends at which point the scripts were enqueued.

### Late Enqueued Scripts
For themes I make or work with, I use a function (much like Poseidon) to register all associated scripts for the theme in the `functions.php` file. The only scripts that get enqueued in this function and at this point are ones I know are used on every page, like maybe some navigation or init scripts.

The rest of the scripts here are only registered and then I only have them enqueued when they're needed. So as an example lets say I have a script that is used for social share buttons on the blog posts. The only place on the website these buttons are used is in the blog posts, so I don't enqueue this unless the current page is a blog post.

The reason I register them early though is that it allows me to enqueue them by simply calling their handle rather than the full file path. It also means I don't have a bunch of scripts loading on pages that don't need them.

The issue is Hyperdrive while hooked into `wp_head`, even with a priority of one, misses these scripts and fails to add them to the `fetchInject()` array and dequeue them. We can get around this by hooking into a different area, but I'll go over that in the next section.

## Hooks
Here I'll try to break down the different types of behaviour I've observed with a few different hooks.

### `wp_head`
This is the current hook used in Hyperdrive. While using this hook Hyperdrive populates the `fetchInject()` array with all `head` scripts and any of the footer scripts enqueued in the footer without a priority to the `wp_enqueue_scripts` hook.

What it doesn't catch is any scripts registered to `wp_enqueue_scripts` but not enqueued until later, it also seems it doesn't catch scripts registered to `wp_enqueue_scripts` with a high priority.

**Pros**
- Catches and dequeues early enqueued footer scripts.
- Doesn't dequeue `head` scripts.

**Cons**
- Catches all enqueued `head` scripts and thus loads them again.
- Fails to catch and dequeue late enqueued footer scripts.
- Fails to catch and dequeue scripts enqueued with a high priority.

### `wp_footer`
Hooking into `wp_footer` seems to result in most footer scripts being caught (added to the `fetchInject()` array) and dequeued, as long as they're not set with a high priority.

It does, however, still catch all `head` scripts.

**Pros**
- Catches and dequeues early enqueued footer scripts.
- Catches and dequeues late enqueued footer scripts.
- Doesn't dequeue `head` scripts.

**Cons**
- Catches all enqueued `head` scripts and thus loads them again.
- Fails to catch and dequeue footer scripts enqueued with a high priority.

### `wp_print_head_scripts`
> If we're doing anything with scripts enqueued in the `head`, it should be done with `wp_print_head_scripts()` instead

This hook doesn't seem to work at all, hooking into it fails to catch or dequeue any scripts, either in the `head` or elsewhere. The documentation for this function talks about `wp_print_footer_scripts()` and I'd probably bet money on this is either deprecated and not used anymore, or it's completely broke.

###  `wp_print_footer_scripts`
Hooking into `wp_print_footer_scripts` seems to result in all footer scripts being caught, even if they're set with a high priority.

It does, however, still catch all `head` scripts.

**Pros**
- Catches early enqueued footer scripts.
- Catches late enqueued footer scripts.
- Catches footer scripts enqueued with a high priority.
- Doesn't dequeue `head` scripts.

**Cons**
- Catches all enqueued `head` scripts and thus loads them again.
- Doesn't dequeue any scripts (that I noticed).

###  `wp_print_scripts`
> `wp_print_scripts()` is used outside what WP calls the "front end"

This is true, although we can get around that by doing a check against `is_admin()` to ensure we only hook into this on the front end.

One caveat of this though is WordPress doesn't class the login pages as being part of the admin, even though they technically are (i.e. `wp-login.php` & `wp-register.php`). It's easy enough to fix this though by creating a function that checks against the `$pagenow` global variable. Something like the following should work:

```
function is_login_page() {
  global $pagenow;
  return in_array($pagenow, array('wp-login.php', 'wp-register.php'));
} 
```

Hooking into `wp_print_scripts` seems to result in all `head` scripts and any footer scripts enqueued early being caught and dequeued, even if they're set with a high priority.

It does, however, fail to catch late enqueued or high priority footer scripts.

**Pros**
- Catches and dequeues early enqueued footer scripts.

**Cons**
- Catches and dequeues all enqueued `head` scripts, including those enqueued with a high priority.
- Fails to catch and dequeue late enqueued footer scripts.
- Fails to catch and dequeue footer scripts enqueued with a high priority set.

### So What Now?
After running these tests, it's clear that all the working hooks (at least those listed above) come with the same _con_ of adding `head` scripts to the `fetchInject()` array. The only hook that does manage to dequeue the scripts added to the `fetchInject()` (including `head` scripts) is `wp_print_scripts`.

So with that being said, it may be we have to utilise multiple hooks or figure out how to detect which scripts are added to the `head` and exclude them from being processed.

## Other Points
> Attempting to override load order of scripts loaded in the HEAD of a document could lead to unexpected issues for `async` scripts loaded early (like Google Analytics), which we should not be concerned about as they're already being loaded asynchronously and typically do not contain dependencies.

Google Analytics and a lot of other `async` scripts shouldn't be too much of a concern. They're added as inline scripts which Hyperdrive ignores. Obviously, this might be a concern when/if WordPress adds the ability to `async` / `defer` scripts, the scripts have been hard coded with `async` / `defer`, or the site has another plugin installed that is adding `async` to said scripts.

> Then let's say we did have a standard way to skip over existing type=async scripts in the `head` (without scraping the DOM), we'd still not be doing justice to pushing for Web standards in existing themes which could be updated and improve the performance for all theme users.

> Beyond that, and until more theme research on scripting patterns is available, my continued suggestion is that we should not attempt to provide compatibility for themes using blocking scripts as doing so will have negative consequences both for theme maintainers and WP Core.

While I agree with these two points, I don't feel the best route is to hope theme maintainers/creators follow Web Standards. If they don't abide by them now, how can we expect them to do so at any point? I know of a lot of theme creators that just spew out new themes to make some quick money and don't provide on-going support or updates.

WordPress's own [theme guidelines](https://make.wordpress.org/themes/handbook/review/) unfortunately don't list anything to do with blocking scripts 👎. I feel that even if we were to ask them to get something added about blocking scripts that they wouldn't (although we should still try). As there are so many themes available through the theme directory it would be hard for them to police it I think, especially considering they don't follow it themselves (with jQuery).

Another thing I feel would be hard to get them to change would be moving jQuery to the footer by default as it would break a lot of sites and plugins.

I feel a better approach would be to add support for fixing blocking scripts in themes and plugins but to as an additional option in the admin UI for Hyperdrive (when it has one). We'd have it set to be default off and would come with a warning for anyone who wants to enable it that it might break their sites if they have themes or plugins that don't follow best practices. 

The same would go for adding jQuery to the `fetchInject()` array, with it being included in the `head` at a core level it probably isn't the best idea to default to dequeuing it. That doesn't mean we can't add it as an option for users who are willing to test it anyway. In my themes, I have no scripts that rely on jQuery in the `head` so dequeuing jQuery would improve site performance for me without breaking anything. That's if there aren't any plugins installed adding scripts to the `head`.

## Final Points
I feel like the best approach would be to at least try and help users with themes and plugins that don't conform to best practices improve their site speed. Even if it means we plaster warnings in the admin UI that their theme or plugins don't follow Web Standards and as a result will be negatively impacting their site performance.

We could group plugin settings by risk factors of how likely they could cause issues with plugins and themes. As long as we make it clear that Hyperdrive will try and fix the issues with the scripts with certain options enabled, it's not guaranteed and that they should test to make sure nothing breaks after enabling "high-risk options", like moving jQuery out of the `head`.

> Create or find Core issue for natively enqueuing `async` scripts in the `head` so Hyperdrive is able to access by API `head` scripts without scraping DOM.

With regards to this, there are a few issues in the WordPress Trac that kind of talk about this, the main one being [this](https://core.trac.wordpress.org/ticket/12009) (at least, that I've found). Unfortunately, the discussion has been on-going for seven years 😮!

If the ability to add `async`, `defer` or any other tags for that matter was added to the `wp_script_add_data()` function then we'd be able to check which scripts had `async` like we check for the conditional tag.

## Main Issues Found
- [ ] `jQuery core` & `jQuery migrate` are always loaded twice.
- [ ] `jQuery migrate` fails to load without the patch in PR #27 
- [ ] Hooking into `wp_head` without a priority causes dequeuing to fail on some scripts.
- [ ] `head` scripts are always included in the `fetchInject()` array.
- [ ] Dequeuing is inconsistent across hooks.

_**Note:** All testing has been carried out on various themes including the default and Poseidon (as that's mentioned in the OP) for both Hyperdrive v1.0.0-beta.0 (The plugin directory version) and the current master v1.0.0-beta.3._

**Comment by jhabdas** [comment]
Excellent, excellent write-up. I'm going to reread this a few more times, mull it over more and do some testing locally with a few different themes before responding.

While conceptualizing Hyperdrive Admin UI recently after watching [this Spaceballs clip](http://www.youtube.com/watch?v=ygE01sOhzz0), here's what I envision:

- Light Speed
- Ridiculous Speed
- Ludicrous Speed
- Plaid

And there will be a lever enabling the user to tune it which will run #17 (and whatever comes after it) based on their chosen theme and perhaps installed plugins. Beyond, and out of empathy for the majority of WordPress users, I'd prefer things to get not much more complex. Though as I said I need to mull it over. And at my top priority right now is getting #9 in place and covering the codebase with tests so we can as they say, floor it.

**Comment by EiraStudios** [comment]
@jhabdas Haha 😆 I love that film.

For Hyperdrive v1.0.0 release, what should it do with jQuery? It's always enqueued into the `head`, so should Hyperdrive ignore adding it to the `fetchInject()` array? If we include processing of jQuery then we'll probably need to include the other `head` scripts to prevent it breaking things.

Obviously, this can be changed in the future, but for the initial release would it be better to exclude jQuery from being processed to ensure better plugin compatibility for coming out of beta?

**Comment by jhabdas** [comment]
I'm restoring this to bug status because in my [original prototype](https://gist.github.com/jhabdas/64e8380010e43a526fb9c9ee511fad17) I was not seeing this behavior. As a result, I believe there is either a bug in the beta, a possible regression or something else which needs to be investigated.

**Comment by jhabdas** [comment]
Discovered root cause in #61, and we only needed 9% unit test coverage to do it.

**Comment by jhabdas** [comment]
Closing in favor of #68. Mental note to loop back and further consider https://github.com/comfusion/hyperdrive/issues/16#issuecomment-301244971 once the root cause of this issue is addressed with consideration for OP and the following public statement made on Medium on May 27: https://medium.com/@jhabdas/hi-ian-hyperdrive-builds-on-top-of-a-single-foundation-which-is-based-on-emerging-web-standards-64f46d0b50cf

**Comment by jhabdas** [comment]
Root cause addressed as discussed here with sample waterfall: https://github.com/comfusion/hyperdrive/issues/7#issuecomment-307549870
